### PR TITLE
UCS/ARBITER/TEST: Add function to push element to group head

### PR DESCRIPTION
**Pull request by yosefe**
Would be used to prioritize RC/DC flow control messages over regular data

### CODE REVIEW

**Comment by yosefe** [comment]
@hoopoepg @brminich can u pls look?

**Comment by hoopoepg** [comment]
only one question: in arbiter.h there is scheme where head of group points to head of another group. I don't see where you update pointers to group head for another groups

File: src/ucs/datastruct/arbiter.h [hunk 1]
```diff
@@ -268,6 +272,18 @@ static inline void ucs_arbiter_group_desched(ucs_arbiter_t *arbiter,
     }
 }
 
+
+/**
+ * Initialize an element object.
+ *
+ * @param [in]  elem    Element to initialize.
+ */
+static inline void ucs_arbiter_elem_init(ucs_arbiter_elem_t *elem)
```
Comment (brminich) [inline src/ucs/datastruct/arbiter.h]: why need to move it?

**Comment by yosefe** [comment]
@hoopoepg it's done by ucs_list_insert_replace()
can you pls redo FC fix over this arbiter API, and test it on scale, before we merge? to have more testing coverage for arbiter code..

**Comment by hoopoepg** [comment]
@yosefe I see
will implement on new API

File: src/ucs/datastruct/arbiter.c [hunk 2]
```diff
@@ -49,6 +50,49 @@ void ucs_arbiter_group_push_elem_always(ucs_arbiter_group_t *group, ucs_arbiter_
...
+
+    head       = tail->next;
+    elem->next = head;        /* Point to first element */
+    tail->next = elem;        /* Point previous element to new one */
+
+    if (head->list.next != NULL) {
+        ucs_assert(arbiter != NULL);
+        ucs_arbiter_group_head_replaced(arbiter, head, elem);
+    } else {
+        elem->list.next = NULL; /* Mark the new head as un-scheduled */
```
Comment (brminich) [inline src/ucs/datastruct/arbiter.c]: minor: seems you can move this init to the beginning of the func and this branch (and line 78) will not be needed

**Comment by hoopoepg** [comment]
@yosefe I implemented FC over this API: https://github.com/openucx/ucx/pull/3443
but it still hangs. it seems we have reorder arbiter groups to set high prio first in pricessing

File: src/ucs/datastruct/arbiter.c [hunk 3]
```diff
@@ -49,6 +50,49 @@ void ucs_arbiter_group_push_elem_always(ucs_arbiter_group_t *group, ucs_arbiter_
...
+        return;
+    }
+
+    head       = tail->next;
+    elem->next = head;        /* Point to first element */
+    tail->next = elem;        /* Point previous element to new one */
+
+    if (head->list.next != NULL) {
+        ucs_assert(arbiter != NULL);
+        ucs_arbiter_group_head_replaced(arbiter, head, elem);
```
Comment (hoopoepg) [inline src/ucs/datastruct/arbiter.c]: replace to
```
if (ucs_list_prev(&head->list, ucs_arbiter_elem_t, list) == head) {
             arbiter->current = elem;
             ucs_list_head_init(&elem->list);
} else {
             ucs_arbiter_group_head_replaced(arbiter, head, elem);
}
```

**Comment by hoopoepg** [comment]
found issue, will update PR tomorrow

**Comment by mellanox-github** [comment]
Test PASSed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/9644/ for details (Mellanox internal link).

**Comment by mellanox-github** [comment]
Test FAILed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/9662/ for details (Mellanox internal link).

**Comment by hoopoepg** [comment]
failed due to git issue

**Comment by hoopoepg** [comment]
bot:mlx:retest

**Comment by mellanox-github** [comment]
Test FAILed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/9673/ for details (Mellanox internal link).

**Comment by brminich** [comment]
bot:mlx:retest

**Comment by mellanox-github** [comment]
Test PASSed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/9679/ for details (Mellanox internal link).

**Pull request merged**